added config path to Service interface

jasmingacic opened the pull request:
For some sort of automations it is practical to get information where service files are stored.

kardianos commented:
This may be alright. But, the ConfigPath would need to be a separate interface; right now it breaks windows I think.

jasmingacic commented:
I haven't tested it with Windows. Do you have a suggestion how to avoid
breaking the build for Windows?

On Thu, Dec 17, 2020, 18:30 Daniel Theophanes <notifications@github.com>
wrote:

> This may be alright. But, the ConfigPath would need to be a separate
> interface; right now it breaks windows I think.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kardianos/service/pull/255#issuecomment-747585110>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB7RER3CBVILCWRFDTOKPA3SVI52HANCNFSM4U7ZSOSA>
> .
>

kardianos commented:
GOOS=windows go build .

jasmingacic commented:
@kardianos Fixed it for windows. 

I also wanted to check with you if I did it right. In most of cases I renamed `configPath() string, error` to `ConfigPath()string` and refactored accordingly. Do you think that's the best approach or maybe it should be done differently?

kardianos commented:
So yes, for this change to happen:

 * Make the ConfigPath return (string error).
 * Remove it from the service interface.
 * Make an optional interface it may support (windows won't implement).
 * Require a type assertion for it to use.
 *

jasmingacic commented:
@kardianos  Do you mind if I call the optional interface Config? (I can't think of a better name right now)

kardianos commented:
No. How about `Metaer` :) or `ConfigPather`

hmmm, maybe `PatholoConfiger` ... I'm getting worse.

Not `Config`, but maybe not `PatholoConfiger` either.

jasmingacic commented:
I think ConfigInfo sound ok then we can have ConfigPath and what other information might be stored there.

jasmingacic commented:
@kardianos I just refactored it to use the optional interface ConfigInfo, later we can decide on the name.

kardianos commented on service.go:
```diff
@@ -387,6 +387,19 @@ type Service interface {
+}
+
+//ConfigPath returns location of the service file where applicable
+func ConfigPath(s Service) string {
```
Remove this helper function

jasmingacic replied:
If I remove it how are we going to retrieve the path?

jasmingacic replied:
I'm doing it this way
```
func TestConfigPath(t *testing.T) {
	p := &program{}
	sc := &service.Config{
		Name: "go_service_test",
	}
	s, err := service.New(p, sc)
	if err != nil {
		t.Fatalf("New err: %s", err)
	}

	_, err = service.ConfigPath(s)
	if err != nil {
		t.Fatal("Failed to fetch or not implemented")
	}
}
```

kardianos commented on service.go:
```diff
@@ -387,6 +387,19 @@ type Service interface {
 }
 
+//ConfigInfo is an optional interface which allows for certain information to be obtained.
+type ConfigInfo interface {
```
rename to ConfigInfoer

kardianos commented on service.go:
```diff
@@ -387,6 +387,19 @@ type Service interface {
 
+//ConfigInfo is an optional interface which allows for certain information to be obtained.
+type ConfigInfo interface {
+	configpath() string
```
`ConfigPath() (string, error)`

jasmingacic replied:
I can't use that ConfigPath as it is already taken in most implementations. Also it doesn't have to be public anyways since it isn't directly called.

kardianos commented on service.go:
```diff
@@ -387,6 +387,19 @@ type Service interface {
 	Status() (Status, error)
 }
 
+//ConfigInfo is an optional interface which allows for certain information to be obtained.
```
Space after `//`, so:
```
// ConfigInfoer ...
```

kardianos reviewed (changes requested):
Ensure you run
`go mod tidy`
and `go fmt`
prior to commit.

jasmingacic commented on linux_test/Makefile:
```diff
@@ -3,7 +3,7 @@ all: sysv systemd upstart openrc clean
 # compile `go test` binary statically
 test:
-	@CGO_ENABLED=0 go test -installsuffix netgo -a -c ..
+	@CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go test -installsuffix netgo -a -c ..
```
@kardianos can you give this a try. When running `make all` on darwin it was failing.

kardianos replied:
It appears to build on my system, but I haven't tested it.

jasmingacic replied:
I ran `linux_test/make all` for the changes I don't have a way to test windows.

kardianos commented on service.go:
```diff
@@ -387,6 +387,19 @@ type Service interface {
 
+// ConfigInfoer is an optional interface which allows for certain information to be obtained.
+type ConfigInfoer interface {
+	configpath() (string, error)
```
Export this as `ConfigPath() (string, error)`

jasmingacic replied:
As requested

kardianos commented on service.go:
```diff
@@ -387,6 +387,19 @@ type Service interface {
+}
+
+//ConfigPath returns location of the service file where applicable
+func ConfigPath(s Service) (string, error) {
```
Remove this helper function.

kardianos commented on service_aix.go:
```diff
@@ -114,6 +114,11 @@ func (s *aixService) template() *template.Template {
 	}
 }
 
+func (s *aixService) configpath() string {
```
fix definition
Also add a type assertion to all service types:
```
var _ ConfigInfoer = &aixService{}
```

jasmingacic commented:
@kardianos can you help me out with the test for ConfigPath how do you want to have it implemented?

kardianos commented:
Go ahead and read the diffs and correct the other OSes. Also, no need to have an exported method and un-exported method that do the same thing. Just use the exported method.

jasmingacic commented:
So getting config value like this is fine?
```
	_, err = s.(service.ConfigInfoer).ConfigPath()
```
Also I'm not sure what you mean by read the diffs and correct the other OSes, I already did that. Unless I'm missing something.

kardianos commented on service_aix.go:
```diff
@@ -114,6 +116,11 @@ func (s *aixService) template() *template.Template {
 	}
 }
 
+func (s *aixService) configpath() string {
```
Wrong name/signature.

jasmingacic replied:
Fixed

kardianos commented on service_freebsd.go:
```diff
@@ -87,6 +89,11 @@ func (s *freebsdService) template() *template.Template {
 	}
 }
 
+func (s *freebsdService) configpath() string {
```
Wrong name/signatre.

jasmingacic replied:
removed

kardianos commented on service_openrc_linux.go:
```diff
@@ -77,6 +79,10 @@ func newOpenRCService(i Interface, platform string, c *Config) (Service, error)
 
 var errNoUserServiceOpenRC = errors.New("user services are not supported on OpenRC")
 
+func (s *openrc) ConfigPath() (string, error) {
```
As I previously stated, don't simply wrap an un-exported method, rename the existing method.

jasmingacic replied:
Done

kardianos commented on service_solaris.go:
```diff
@@ -98,6 +100,11 @@ func (s *solarisService) template() *template.Template {
 	}
 }
 
+func (s *solarisService) configpath() string {
```
Wrong name/signature.

jasmingacic replied:
removed

kardianos commented on service_systemd_linux.go:
```diff
@@ -67,6 +69,10 @@ func (s *systemd) Platform() string {
 }
 
+func (s *systemd) ConfigPath() (string, error) {
+	return s.configPath()
```
As I previously mentioned, don't just wrap a private method, rename the existing one.

jasmingacic replied:
Done

kardianos commented on service_upstart_linux.go:
```diff
@@ -61,6 +63,10 @@ func (s *upstart) Platform() string {
 // Upstart will be replaced by systemd in most cases anyway.
 var errNoUserServiceUpstart = errors.New("User services are not supported on Upstart.")
 
+func (s *upstart) ConfigPath() (string, error) {
```
As I previously mentioned, don't just wrap an existing private method, rename it to export it.

jasmingacic replied:
Done

kardianos reviewed (changes requested):
Please read all my comments carefully this time. Please read your diffs carefully.

jasmingacic commented on service_test.go:
```diff
@@ -41,6 +41,22 @@ func TestRunInterrupt(t *testing.T) {
+		t.Fatalf("New err: %s", err)
+	}
+
+	_, err = s.(service.ConfigInfoer).ConfigPath()
```
Is this test valid?

jasmingacic commented:
ping
